lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class ClientError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config)
+      @config = config
+      @modified_options = {}
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+    end
+
+    attr_reader :config
+    attr_reader :scheduled_task, :promise # for specs
+
+    def start
+      return unless config.central_config?
+
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @promise =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @scheduled_task&.cancel
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 300..399
+        resp
+      when 400..499
+        raise ClientError, resp
+      when 500..599
```

Comment:
Maybe you want to make these Ranges constants with names describing each scenario so new Ranges are not created each time this method is called? Could be slightly more efficient.

Reply:
Update: I checked `Concurrent::Promise` and it saves `fetch_config` as a proc so the only benefit of using constants would be for descriptiveness, not efficiency.

Reply:
Good call to check!

I went a bit back and forth of whether I wanted to be more explicit with the status codes we expect. We know what apm server might return. We are already explicit further down with 304 Not Changed. Idk. This is probably fine as is.